File: mvp/src/jungle.js
// 生态 Jungle / Amen 切片地址。
//
// stepIndex 决定原始 break 的切片位置；pitchBranchId 只决定移调。
// 五枝不再伪装成 kick/snare/hat 角色，也不会触发程序化鼓。音频层按这里给出的
// 音频层用颗粒重采样读取真实 WAV：tempoRate 只推进源时间轴，
// pitchRate 只改变颗粒内移调，因而五枝读取同样的 Amen 拍长并占满同一步。

export const JUNGLE_PITCH_SEMITONES = Object.freeze([-7, -3, 0, 3, 7]);

export const JUNGLE_PITCH_LABELS = Object.freeze([
  '−7st', '−3st', '原调', '+3st', '+7st',
]);

const clamp01 = (value) => Math.max(0, Math.min(1, Number(value) || 0));

/**
 * 16 个世界时间格均匀读取 dnber 的 32-slice Amen 网格。最后一格停在 29，
 * 为最高移调保留足够源采样，避免靠近文件尾时被迫缩短。
 */
export function jungleSliceForCell({
  pitchBranchId = 0,
  roleId,
  stepIndex = 0,
  tension = 0.3,
  masterBpm = 60,
  tempoMultiplier = 2,
  amenDurationSeconds = 2.742857142857143,
  amenNativeBeats = 8,
} = {}) {
  const requestedPitch = Number.isFinite(Number(pitchBranchId)) ? Number(pitchBranchId) : Number(roleId);
  const pitchIndex = Math.max(0, Math.min(
    JUNGLE_PITCH_SEMITONES.length - 1,
    Math.trunc(Number.isFinite(requestedPitch) ? requestedPitch : 0),
  ));
  const step = Math.max(0, Math.trunc(Number.isFinite(Number(stepIndex)) ? Number(stepIndex) : 0));
  const semitones = JUNGLE_PITCH_SEMITONES[pitchIndex];
  const jungleBpm = Math.max(1, Number(masterBpm) || 60)
    * Math.max(1, Number(tempoMultiplier) || 2);
  const outputSeconds = 60 / jungleBpm;
  const nativeBeatSeconds = Math.max(0.001, Number(amenDurationSeconds) || 2.742857142857143)
    / Math.max(1, Number(amenNativeBeats) || 8);
  const pitchRate = 2 ** (semitones / 12);
  const tempoRate = nativeBeatSeconds / outputSeconds;
  return {
    amenStep: Math.min(29, (step % 16) * 2),
    pitchIndex,
    semitones,
    jungleBpm,
    nativeBeatSeconds,
    tempoRate,
    pitchRate,
    // 保留字段名给音频节点；只含移调，tempo 由颗粒源位置独立推进。
    playbackRate: pitchRate,
    outputSeconds,
    velocity: 0.92 + clamp01(tension) * 0.08,
  };
}

/**
 * 将一个固定输出时值的 chop 拆成交叉颗粒。每粒的输出起点按
 * tempoRate 映射到 Amen 源时间轴，而粒内 playbackRate 只使用 pitchRate。
 * 这是 Web Audio 原生节点下将“移调”与“整片时值”分离的最小实现。
 */
export function jungleGrainPlan(slice, {
  grainSeconds = 0.1,
  overlap = 0.5,
} = {}) {
  const total = Math.max(0.001, Number(slice?.outputSeconds) || 0.5);
  const grain = Math.max(0.02, Math.min(total, Number(grainSeconds) || 0.1));
  const overlapRatio = Math.max(0.1, Math.min(0.8, Number(overlap) || 0.5));
  const hop = grain * (1 - overlapRatio);
  const tempoRate = Math.max(0.001, Number(slice?.tempoRate) || 1);
  const pitchRate = Math.max(0.001, Number(slice?.pitchRate) || 1);
  const grains = [];
  for (let outputOffset = 0; outputOffset < total - 1e-6; outputOffset += hop) {
    const outputDuration = Math.min(grain, total - outputOffset);
    grains.push({
      outputOffset,
      outputDuration,
      sourceOffset: outputOffset * tempoRate,
      sourceTimelineDuration: outputDuration * tempoRate,
      sourceDuration: outputDuration * pitchRate,
      playbackRate: pitchRate,
    });
  }
  return grains;
}

// 配置/评分字段仍叫 roleDiversity 以保持存档与 LLM 契约兼容；实际语义已是
// “同一天用到多少个移调枝”。
export function jungleRoleDiversity(cells = [], pitchCount = JUNGLE_PITCH_SEMITONES.length) {
  const pitches = new Set((cells ?? [])
    .map((cell) => Number(cell?.pitchBranchId))
    .filter((pitch) => Number.isInteger(pitch) && pitch >= 0 && pitch < pitchCount));
  const onsets = new Set((cells ?? [])
    .map((cell) => Number(cell?.stepIndex))
    .filter(Number.isInteger));
  if (!onsets.size) return 0;
  return pitches.size / Math.min(pitchCount, onsets.size);
}

File: flock-voice-engine/server/audio_worker/jungle.py
"""Frozen Python port of the browser Jungle address/grain planner."""
from __future__ import annotations

import math
from typing import Any

JUNGLE_PITCH_SEMITONES = (-7, -3, 0, 3, 7)
AMEN_TRANSIENT_STEPS = (0, 2, 4, 6, 7, 8, 10, 12, 14, 15, 16, 18, 20, 22, 24, 28)


def _number(value: Any, fallback: float) -> float:
    try:
        result = float(value)
    except (TypeError, ValueError):
        return fallback
    return result if math.isfinite(result) else fallback


def reverse_amen_offset(duration: float, forward_offset: float, source_duration: float) -> float:
    total = max(0.001, _number(duration, 0.001))
    end = (_number(forward_offset, 0.0) + max(0.0, _number(source_duration, 0.0))) % total
    return (total - end) % total


def jungle_edit_plan(day=0, tension=0, onset_count=0, conflict_ratio=0,
                     pattern_similarity=0) -> dict[str, Any]:
    dense = _number(onset_count, 0) >= 8
    conflict = _number(conflict_ratio, 0) >= 0.05
    stale = _number(pattern_similarity, 0) >= 0.82
    t = _number(tension, 0)
    break_edit = "dropout" if dense and conflict else "repeat4" if stale and t >= .7 else \
        "repeat2" if stale or (dense and t >= .5) else "hold"
    tone_edit = "filter" if conflict else "reverse" if stale and t >= .75 else \
        "crush" if dense and t >= .6 else "dub" if int(_number(day, 0)) % 4 == 0 and t >= .4 else "clean"
    return {"breakEdit": break_edit, "toneEdit": tone_edit,
            "evidence": {"onsetCount": _number(onset_count, 0),
                         "conflictRatio": _number(conflict_ratio, 0),
                         "patternSimilarity": _number(pattern_similarity, 0), "tension": t}}


def jungle_slice_for_cell(pitch_branch_id=0, role_id=None, step_index=0, tension=.3,
                          master_bpm=60, tempo_multiplier=2,
                          amen_duration_seconds=2.742857142857143, amen_native_beats=8,
                          **aliases) -> dict[str, Any]:
    pitch_branch_id = aliases.get("pitchBranchId", pitch_branch_id)
    role_id = aliases.get("roleId", role_id)
    step_index = aliases.get("stepIndex", step_index)
    tension = aliases.get("tension", tension)
    master_bpm = aliases.get("masterBpm", master_bpm)
    tempo_multiplier = aliases.get("tempoMultiplier", tempo_multiplier)
    amen_duration_seconds = aliases.get("amenDurationSeconds", amen_duration_seconds)
    amen_native_beats = aliases.get("amenNativeBeats", amen_native_beats)
    requested = _number(pitch_branch_id, _number(role_id, 0))
    pitch_index = max(0, min(len(JUNGLE_PITCH_SEMITONES) - 1, math.trunc(requested)))
    step = max(0, math.trunc(_number(step_index, 0)))
    semitones = JUNGLE_PITCH_SEMITONES[pitch_index]
    jungle_bpm = max(1, _number(master_bpm, 60)) * max(1, _number(tempo_multiplier, 2))
    output_seconds = 60 / jungle_bpm
    native_beat_seconds = max(.001, _number(amen_duration_seconds, 2.742857142857143)) / max(
        1, _number(amen_native_beats, 8))
    pitch_rate = 2 ** (semitones / 12)
    tempo_rate = native_beat_seconds / output_seconds
    return {"amenStep": AMEN_TRANSIENT_STEPS[step % len(AMEN_TRANSIENT_STEPS)],
            "pitchIndex": pitch_index, "semitones": semitones, "jungleBpm": jungle_bpm,
            "nativeBeatSeconds": native_beat_seconds, "tempoRate": tempo_rate,
            "pitchRate": pitch_rate, "playbackRate": pitch_rate, "outputSeconds": output_seconds,
            "velocity": .92 + max(0, min(1, _number(tension, 0))) * .08}


def jungle_grain_plan(slice_value: dict[str, Any], grain_seconds=.1, overlap=.5, **aliases) -> list[dict[str, float]]:
    grain_seconds = aliases.get("grainSeconds", grain_seconds)
    total = max(.001, _number(slice_value.get("outputSeconds"), .5))
    grain = max(.02, min(total, _number(grain_seconds, .1)))
    overlap_ratio = max(.1, min(.8, _number(overlap, .5)))
    hop = grain * (1 - overlap_ratio)
    tempo_rate = max(.001, _number(slice_value.get("tempoRate"), 1))
    pitch_rate = max(.001, _number(slice_value.get("pitchRate"), 1))
    output = []
    offset = 0.0
    while offset < total - 1e-6:
        duration = min(grain, total - offset)
        output.append({"outputOffset": offset, "outputDuration": duration,
                       "sourceOffset": offset * tempo_rate,
                       "sourceTimelineDuration": duration * tempo_rate,
                       "sourceDuration": duration * pitch_rate, "playbackRate": pitch_rate})
        offset += hop
    return output
